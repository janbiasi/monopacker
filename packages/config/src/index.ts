export { Config, ResolverConfig, PackConfig, PackageManager } from './types';
export { getConfig } from './resolve';
export { validateConfig } from './validator';
export { defaultFilesToCopy, defaultIgnorePatterns } from './defaults';
