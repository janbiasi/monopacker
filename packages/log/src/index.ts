export * from './default';
export { LogLevel, LoggableLevel, loggableLevels } from './level';
export { createLogger, getLogStack } from './logger';
export { Logger } from './types';
export { Namespace, namespaces } from './namespace';
