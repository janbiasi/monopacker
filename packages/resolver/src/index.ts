export { Graph, GraphResolutionEntry } from './types/graph';
export { PackageJson, PackageDependencyMap, PackagePeerDependenciesMeta } from './types/package';
export { TargetGraph } from './types/target-graph';

export { getPackageGraph, getPackageResolutions } from './graph';
export { parsePackage, parsePackages } from './package';
export { resolveTargetGraph, resolveTargetPackageName } from './resolve';
export { searchPackages, searchFiles } from './search';
