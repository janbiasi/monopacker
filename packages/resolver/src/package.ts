import type { PackageJson } from './types/package';

export const parsePackage = (absolutePath: string): { pkg: PackageJson, location: string } => {
	try {
		return {
			pkg: require(absolutePath),
			location: absolutePath,
		}
	} catch (err) {
		throw new Error(`Unable to load package.json at ${absolutePath}: ${err}`);
	}
};

export const parsePackages = (paths: string[]) => {
	return paths.map(parsePackage);
};
