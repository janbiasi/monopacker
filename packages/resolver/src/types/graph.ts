import type { PackageDependencyMap } from './package';

export interface GraphResolutionEntry {
	internal: PackageDependencyMap;
	external: string[];
	remote: PackageDependencyMap;
	peer: PackageDependencyMap;
}

/**
 * The main graph aggregarted by the resolver. This is the base
 * of all actions ran by the packing process.
 */
export interface Graph {
	local: {
		[packageName: string]: {
			path: string;
			version: string;
		};
	};
	resolution: {
		[packageName: string]: GraphResolutionEntry;
	};
}
