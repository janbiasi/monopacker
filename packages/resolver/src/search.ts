import globby from 'globby';
import { Config, defaultIgnorePatterns } from '@monopacker/config';

/**
 * Find all packages in the given directory.
 * @param {string} rootDir - The directory to search.
 * @param {SearchOptions} options - Options for finding packages.
 * @returns
 */
export const searchPackages = async (rootDir: string, config: Config['resolver']) => {
	const packages = await globby(['**/package.json'], {
		cwd: rootDir,
		ignore: config?.ignore || defaultIgnorePatterns,
		followSymbolicLinks: config?.followSymbolicLinks === false ? false : true,
		deep: config?.deep || 7
	});

	if (packages.length === 0) {
		throw new Error(`Unable to find any packages in ${rootDir}`);
	}

	return packages;
};

/**
 * Basic search via globs
 * @param rootDir
 * @param globs
 * @returns
 */
export async function searchFiles(rootDir: string, globs: string[]) {
	try {
		return await globby(globs, {
			cwd: rootDir,
			deep: 20
		});
	} catch (err) {
		throw new Error(`Failed to copy files "${globs.join(',')}" in ${rootDir}:\n\n${err}`);
	}
}
