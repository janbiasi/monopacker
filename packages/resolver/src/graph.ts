import type { ResolverConfig } from '@monopacker/config';
import type { Graph, GraphResolutionEntry } from './types/graph';
import type { PackageJson } from './types/package';
import { parsePackages } from './package';
import { searchPackages } from './search';
import { resolve } from 'path';

/**
 * Aggregate resolutions of dependencies for a single package, based on a list of other internal packages.
 * TODO: Respect semver ranges for each internal package
 * @param pkg
 * @param internalPkgNames
 * @returns
 * @private
 */
export function getPackageResolutions(
	pkg: PackageJson,
	internalPkgNames: string[],
	config: ResolverConfig = {}
): GraphResolutionEntry {
	const dependencies = pkg.dependencies || {};
	const peerDependencies = pkg.peerDependencies || {};
	const explicitExternals = config.externals || [];
	const includePeers = config.includePeers === false ? false : true;

	const resolutions: GraphResolutionEntry = {
		internal: {},
		external: [],
		remote: {},
		peer: {},
	};

	for (const dependency in dependencies) {
		if (explicitExternals.includes(dependency)) {
			resolutions.external.push(dependency);
			continue;
		}

		if (internalPkgNames.includes(dependency)) {
			// internal dependency, should be packed locally
			resolutions.internal[dependency] = dependencies[dependency];
		} else {
			// normal dependency, set as remote
			resolutions.remote[dependency] = dependencies[dependency];
		}
	}

	for (const peerDependency in peerDependencies) {
		resolutions.peer[peerDependency] = peerDependencies[peerDependency];

		if (explicitExternals.includes(peerDependency)) {
			resolutions.external.push(peerDependency);
			continue;
		}

		if (internalPkgNames.includes(peerDependency)) {
			// internal peers should always be packed, even if not explicitly specified via `config.resolver.includePeers`
			resolutions.internal[peerDependency] = peerDependencies[peerDependency];
		} else {
			// other peers should be defined as normal prod dependency
			// TODO: this should not happen, this would be a developer's mistake
			if (!resolutions.remote[peerDependency] && includePeers) {
				resolutions.remote[peerDependency] = peerDependencies[peerDependency];
			}
		}
	}

	return resolutions;
}

/**
 * Get the main graph for all packages in the repository
 * @param rootDir
 * @param config
 * @returns
 */
export async function getPackageGraph(rootDir: string, config: ResolverConfig = {}): Promise<Graph> {
	const packagePaths = await searchPackages(rootDir, config);
	const packages = parsePackages(packagePaths.map((packagePath) => resolve(rootDir, packagePath)));
	const graph: Graph = {
		local: {},
		resolution: {},
	};

	// 1. build local graph to know which packages exist locally
	for (const { pkg, location } of packages) {
		if (graph.local[pkg.name]) {
			throw new Error(
				`Duplicate package "${pkg.name}" detected at ${location}, unable to resolve internals correctly`
			);
		}

		graph.local[pkg.name] = {
			version: pkg.version,
			path: location,
		};
	}

	const internalPackageNames = Object.keys(graph.local);

	// 2. build resolution graph based on known package
	for (const { pkg } of packages) {
		graph.resolution[pkg.name] = getPackageResolutions(pkg, internalPackageNames, config);
	}

	return graph;
}
