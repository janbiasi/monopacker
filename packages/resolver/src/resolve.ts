import { logger } from './logger';
import { Graph } from './types/graph';
import { PackageJson, PackageDependencyMap } from './types/package';
import { TargetGraph } from './types/target-graph';

export function resolveTargetPackageName(target: string, localPackageNames: string[]) {
	let targetPackageName: string | undefined;

	if (localPackageNames.includes(target)) {
		targetPackageName = target;
	} else {
		try {
			const targetPackage: PackageJson = require(target);
			if (targetPackage && targetPackage.name) {
				targetPackageName = targetPackage.name;
			}
		} catch (err) {
			// Silently ignore, we'll check if the target is a local package later
		}
	}

	return targetPackageName;
}

/**
 * We want to know every internal dependency even if they are nested deeply in N packages plus we also
 * want to abort further processing if we detect unexpected cyclic dependencies.
 *
 * Example resolve cases:
 * - App -> A + B should resolve  A and B
 * - App -> A + B -> C should resolve A, B and C
 * - App -> A -> B -> C should resolve A, B and C
 * - App -> A -> B -> C -> A should throw an error (cyclic dependency)
 *
 * @param {string[]} currentResolutionPath - The package resolution graph which represents the dependency chain
 * @param {PackageDependencyMap} dependenciesToResolve - The dependencies which we want to resolve
 * @param {PackageDependencyMap} destinationTree - The final dependency tree which we want to populate
 * @param {Graph} mainRepoGraph - The main package resolution graph of the current project
 *
 * @returns {string[], PackageDependencyMap} - The final resolution chain and the destination tree
 */
function resolveDeepInternalDependencies(
	currentResolutionPath: string[],
	dependenciesToResolve: PackageDependencyMap,
	destinationTree: PackageDependencyMap,
	mainRepoGraph: Graph
): [string[], PackageDependencyMap] {
	for (const dependency in dependenciesToResolve) {
		const currentInternalPeerDependencies = mainRepoGraph.resolution[dependency].internal;

		/**
		 * If we detect that a dependency is already in the target tree, we know that there's
		 * something wrong with the source package graphs and we should stop further execution
		 * anyway instead of doing some "black box magic"
		 */
		if (destinationTree.hasOwnProperty(dependency)) {
			throw new Error(
				`Cycling dependency "${dependency}" detected (resolution: ${currentResolutionPath.join(' -> ')})`
			);
		}

		logger.debug(
			`Resolving internal dependency "${dependency}" (resolution: ${currentResolutionPath.join(' -> ')})`
		);

		// always patch current dependency into destination tree (direct resolution)
		destinationTree[dependency] = dependenciesToResolve[dependency];

		if (Object.keys(currentInternalPeerDependencies).length === 0) {
			// no peer internals found for currentd dependency, skipping
			logger.debug(`No peer internals found for "${dependency}" (parent: ${currentResolutionPath.join(' -> ')})`);
			continue;
		}

		logger.debug(`Found peer internals for "${dependency}" (parent: ${currentResolutionPath.join(' -> ')})`);

		const [recursiveResolutionChain, recursiveDependencies] = resolveDeepInternalDependencies(
			[...currentResolutionPath, dependency],
			currentInternalPeerDependencies,
			destinationTree,
			mainRepoGraph
		);

		// TODO: maybe we need to check for parent scope compatibility issues when resolving circular dependencies

		currentResolutionPath.push(`${recursiveResolutionChain.join(' -> ')}`);
		destinationTree = {
			...destinationTree,
			...recursiveDependencies,
		};
	}

	return [currentResolutionPath, destinationTree];
}

/**
 * Get the target resolution for a certain application within the graph.
 * @param {Graph} graph - The current repository graph
 * @param {string}Â target - Target package name or path to the app which should get packed
 * @returns {TargetGraph} - The target resolution graph
 */
export function resolveTargetGraph(graph: Graph, target: string): TargetGraph {
	const localPackageNames = Object.keys(graph.local);
	const targetPackageName = resolveTargetPackageName(target, localPackageNames);

	if (!targetPackageName) {
		throw new Error(`Could not resolve target package "${target}", available are: ${localPackageNames.join(', ')}`);
	}

	if (!graph.resolution[targetPackageName] || Object.keys(graph.resolution[targetPackageName]).length === 0) {
		throw new Error(`Target package "${target}" does not have any dependencies, monopacker will have no effect`);
	}

	const [resolution, internals] = resolveDeepInternalDependencies(
		[targetPackageName],
		graph.resolution[targetPackageName].internal,
		{},
		graph
	);

	return {
		resolution,
		internals,
		name: targetPackageName,
	};
}
