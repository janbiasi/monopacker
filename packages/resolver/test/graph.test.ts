import type { PackageJson } from "../src/types/package";
import { getPackageGraph, getPackageResolutions } from "../src/graph";
import { resolve } from "path";

const FIXTURES_DIR = resolve(__dirname, 'fixtures');
const getFixturePath = (name: string) => resolve(FIXTURES_DIR, name);

const testPackage: PackageJson = {
	name: 'test',
	version: '1.0.0',
	dependencies: {
		react: '17.0.2',
        internalPackage: '^2.1.0'
	},
	devDependencies: {
		'react-dom': '17.0.2'
	},
	peerDependencies: {
		uuid: '3.1.0',
        internalPackage2: '^1.0.1'
	}
};

describe('graph', () => {
    it('should aggregate dependency resolutions with internals and peer internals correctly', () => {
        const resolutions = getPackageResolutions(testPackage, ['internalPackage', 'internalPackage2'])

        expect(resolutions).toMatchSnapshot();
        expect(resolutions.internal).toStrictEqual({
           internalPackage: '^2.1.0',
           internalPackage2: '^1.0.1',
        });
        expect(resolutions.remote).toStrictEqual({
           react: '17.0.2',
           uuid: '3.1.0'
        });
    });

    it('should aggregate dependency resolutions without "includePeers" flag correctly', () => {
        const resolutions = getPackageResolutions(testPackage, ['internalPackage', 'internalPackage2'], {
            includePeers: false
        });

        expect(resolutions).toMatchSnapshot();
        expect(resolutions.internal).toStrictEqual({
           internalPackage: '^2.1.0',
           internalPackage2: '^1.0.1',
        });
        expect(resolutions.remote).toStrictEqual({
           react: '17.0.2'
        });
    });

    it('should aggregate dependency resolutions without internals correctly', () => {
        const resolutions = getPackageResolutions(testPackage, []);
        
        expect(resolutions).toMatchSnapshot();
        expect(resolutions.internal).toStrictEqual({});
        expect(resolutions.remote).toStrictEqual({
           react: '17.0.2',
           internalPackage: '^2.1.0',
           internalPackage2: '^1.0.1',
           uuid: '3.1.0',
        });
    });

    describe('getPackageGraph', () => {
        it("should fail if duplicate internals were detected", () => {
            expect(true).toBe(true);
        });

        describe('fixture: alpha', () => {
            const fixtureDir = getFixturePath('project-alpha');

            it('should resolve dependencies correctly', async () => {
                const graph = await getPackageGraph(fixtureDir);
                expect(Object.keys(graph.local).sort()).toMatchSnapshot();
                expect(graph.resolution).toMatchSnapshot();
            });
        
        })
    });
});