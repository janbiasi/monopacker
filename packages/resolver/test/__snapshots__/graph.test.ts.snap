// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graph getPackageGraph fixture: alpha should resolve dependencies correctly 1`] = `
Array [
  "@project-alpha/app",
  "@project-alpha/package-a",
  "@project-alpha/package-b",
]
`;

exports[`graph getPackageGraph fixture: alpha should resolve dependencies correctly 2`] = `
Object {
  "@project-alpha/app": Object {
    "external": Array [],
    "internal": Object {
      "@project-alpha/package-a": "1.0.0",
      "@project-alpha/package-b": "1.0.0",
    },
    "peer": Object {},
    "remote": Object {
      "react": "17.0.2",
    },
  },
  "@project-alpha/package-a": Object {
    "external": Array [],
    "internal": Object {},
    "peer": Object {},
    "remote": Object {
      "dependency-of-a": "1.0.0",
    },
  },
  "@project-alpha/package-b": Object {
    "external": Array [],
    "internal": Object {},
    "peer": Object {},
    "remote": Object {
      "dependency-of-b": "1.0.0",
    },
  },
}
`;

exports[`graph getPackageGraph fixture: beta should resolve dependencies correctly 1`] = `
Array [
  "@project/app",
  "@project/package-a",
  "@project/package-b",
]
`;

exports[`graph getPackageGraph fixture: beta should resolve dependencies correctly 2`] = `
Object {
  "@project/app": Object {
    "external": Array [],
    "internal": Object {
      "@project/package-a": "1.0.0",
    },
    "peer": Object {},
    "remote": Object {},
  },
  "@project/package-a": Object {
    "external": Array [],
    "internal": Object {
      "@project/package-b": "1.0.0",
    },
    "peer": Object {},
    "remote": Object {},
  },
  "@project/package-b": Object {
    "external": Array [],
    "internal": Object {},
    "peer": Object {},
    "remote": Object {
      "another-package": "1.0.0",
    },
  },
}
`;

exports[`graph getPackageGraph fixture: externals should resolve dependencies correctly with provided externals 1`] = `
Array [
  "@project/app",
  "@project/package-a",
  "@project/package-b",
]
`;

exports[`graph getPackageGraph fixture: externals should resolve dependencies correctly with provided externals 2`] = `
Object {
  "@project/app": Object {
    "external": Array [],
    "internal": Object {
      "@project/package-a": "1.0.0",
    },
    "peer": Object {},
    "remote": Object {},
  },
  "@project/package-a": Object {
    "external": Array [],
    "internal": Object {
      "@project/package-b": "1.0.0",
    },
    "peer": Object {},
    "remote": Object {},
  },
  "@project/package-b": Object {
    "external": Array [],
    "internal": Object {},
    "peer": Object {},
    "remote": Object {
      "another-package": "1.0.0",
    },
  },
}
`;

exports[`graph should aggregate dependency resolutions with internals and peer internals correctly 1`] = `
Object {
  "external": Array [],
  "internal": Object {
    "internalPackage": "^2.1.0",
    "internalPackage2": "^1.0.1",
  },
  "peer": Object {
    "internalPackage2": "^1.0.1",
    "uuid": "3.1.0",
  },
  "remote": Object {
    "react": "17.0.2",
    "uuid": "3.1.0",
  },
}
`;

exports[`graph should aggregate dependency resolutions without "includePeers" flag correctly 1`] = `
Object {
  "external": Array [],
  "internal": Object {
    "internalPackage": "^2.1.0",
    "internalPackage2": "^1.0.1",
  },
  "peer": Object {
    "internalPackage2": "^1.0.1",
    "uuid": "3.1.0",
  },
  "remote": Object {
    "react": "17.0.2",
  },
}
`;

exports[`graph should aggregate dependency resolutions without internals correctly 1`] = `
Object {
  "external": Array [],
  "internal": Object {},
  "peer": Object {
    "internalPackage2": "^1.0.1",
    "uuid": "3.1.0",
  },
  "remote": Object {
    "internalPackage": "^2.1.0",
    "internalPackage2": "^1.0.1",
    "react": "17.0.2",
    "uuid": "3.1.0",
  },
}
`;
